File: leetcode-04/solutions/cpp/0116.cpp
class Solution {
 public:
  Node* connect(Node* root) {
    Node* node = root;

    while (node && node->left) {
      Node* next = node->left;
      while (node) {
        node->left->next = node->right;
        node->right->next = node->next ? node->next->left : NULL;
        node = node->next;
      }
      node = next;
    }

    return root;
  }
};

File: leetcode-04/solutions/python3/0116.py
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        node = root

        while node and node.left:
            next = node.left
            while node:
                node.left.next = node.right
                node.right.next = node.next and node.next.left
                node = node.next
            node = next

        return root
